[24813] Refactor single view edit context

oliverguenther opened the pull request:
This PR re-uses the edit context / edit form functionality built for the table to re-implement the single view edit fields

**TODOS**

- [x] Strip all editing functionality from `edit-field.directive`, it should only handle the activation/deactivation of edit fields.
- [x] Create an edit context for the single view
- [x] Remove `wp-display-attr.directive` since we can manually render fields now on demand.
- [x] Remove `wp-edit-form.directive` and replace with a simpler directive to only find all potential fields (in the case of create-all mode).
- [x] Remove the EditModeState global service. **This removes edit-all functionality for existing WPs**, it was discussed that this is no longer needed.
- [x] Fix tests, lots of them will break due to the removed edit-all
- [x] Remove duplicated `TableWorkPackageField` and `WorkPackageField`
- [x] Find a way to propagate the click area for input fields. This is currently located in the table-only edit-cell handler, and needs to be extracted.

I also took a stab at the query-history spec, which appears to be failing more often than not.

https://community.openproject.com/projects/openproject/work_packages/24813

oliverguenther committed:
Implement single view context for previous edit fields

oliverguenther committed:
Replace wpEditModeState and wpEditForm with one directive

oliverguenther committed:
Fix tests

oliverguenther committed:
Consolidate existing WorkPackageField and TableWorkPackageField

oliverguenther committed:
Fix tests

oliverguenther committed:
Remove single-wp edit in context menu

oliverguenther committed:
Fix more tests

oliverguenther committed:
Extract click position mapper for input fields for re-use

oliverguenther committed:
Test more explicitly in query_history spec

oliverguenther committed:
Always update the edit context

Otherwise, starting to edit in the table and then switching to the
single view will try to activate fields in the wrong edit context.

ulferts commented on frontend/app/components/context-menus/wp-context-menu/wp-context-menu.controller.ts:
```diff
@@ -117,12 +117,6 @@ function wpContextMenuController($scope:any,
-      workPackageId: selected[0].id
-    };
-
-    $state.transitionTo('work-packages.show.edit', params);
```
Doing things like this I believe we should rename the context menu entries to "Bulk move" when more than one wp is selected. It becomes harder and harder to understand where I will end up and why menu items disappear when I have only one/more than one wp selected. 

Until then, we might want to disable the "edit" menu entry.

oliverguenther replied:
I agree, and have changed the bulk actions to be explicit (Bulk move, edit, copy, delete).

ulferts replied:
Thanks :bow:

ulferts commented on frontend/app/components/wp-edit-form/display-field-renderer.ts:
```diff
@@ -0,0 +1,87 @@
+    }
+
+    return name;
+  }
```
Shouldn't we by now be able to just use what the api returns to us which would be a 'date' attribute for milestone-typed wps and start/due-date for the others?

oliverguenther replied:
I've extracted that too far. _It is_ needed in the table, since we need to map start/due date for milestones there, but not in the single view context. I've removed that change so that it is only used in the CellBuilder

ulferts reviewed (changes requested):
### General

![image](https://user-images.githubusercontent.com/617519/28260543-d4d19116-6adb-11e7-81aa-2552f0bca00e.png)
What's not to like about this. 

As a personal preference, I'd like to completely get rid of the full edit mode and believe it could be done just like we do it for the table inline create. Upon creation, the focus could simply jump into the subject input which needs to be filled out in any case. The rest of the fields could be in read mode.

### Backend update bug

This might be non unique to this PR and looks like it needs to be fixed anyway: Changing the work package type from a type having a custom field to one not having the custom field leads to 500:

```
NoMethodError (undefined method `custom_field_115=' for #<WorkPackage:0x007fa0836a2f68>
Did you mean?  custom_fields=):
  
lib/plugins/acts_as_customizable/lib/acts_as_customizable.rb:240:in `method_missing'
lib/api/v3/work_packages/work_packages_shared_helpers.rb:40:in `merge_hash_into_work_package!'
lib/api/v3/work_packages/work_packages_shared_helpers.rb:51:in `write_work_package_attributes'
lib/api/v3/work_packages/work_packages_shared_helpers.rb:62:in `create_work_package_form'
lib/api/v3/work_packages/update_form_api.rb:39:in `block (2 levels) in <class:UpdateFormAPI>'
app/middleware/reset_current_user.rb:47:in `call'
```

### Layouting:

The subject is cut off to the right and the type selector is too small on wp create.

![image](https://user-images.githubusercontent.com/617519/28259818-717dda1e-6ad8-11e7-95e6-348a40f00b03.png)

Placeholder is not italic:
![image](https://user-images.githubusercontent.com/617519/28260002-62fe95cc-6ad9-11e7-9aed-9b74153d9c2d.png)

The 'more' drop down (or rather drop 'up') is cut of: 
![image](https://user-images.githubusercontent.com/617519/28260466-76494a80-6adb-11e7-82ce-d5e2ea9f4259.png)
I imagine this getting worse for languages like german. We might be able to simply align it left.  


### UX

When using inline create, after every entry into a field, all input fields are redraw which leads to them visually disappearing for some time. 

Empty drop downs do not have the '-' (dash) displayed but have an empty option although when opening the drop down, the dash is selected:
![image](https://user-images.githubusercontent.com/617519/28260066-b23d0d30-6ad9-11e7-849c-3a64767c9d8c.png)

### Accessibility
Tabbing through the wp attributes, chromevox reads out too much and too little. It omits the attribute name and reads out the role two times (e.g. for the assignee:
"User lastname, User firstname Button Button"


### JS console:

Switching type on wp create will sometimes cause a lot of js errors being reported in the console:
![image](https://user-images.githubusercontent.com/617519/28259888-c70e9df6-6ad8-11e7-8ef6-e873ead6d875.png)

oliverguenther committed:
Replace container instead of emptying first

oliverguenther committed:
Fix selector on type|subject split

oliverguenther committed:
Refresh open fields instead of redrawing all content

oliverguenther committed:
Remove focus directive on all fields

oliverguenther committed:
Add separate names to bulk actions

oliverguenther committed:
Avoid normalization of date fields outside table

oliverguenther commented:
> The subject is cut off to the right and the type selector is too small on wp create.

> Placeholder is not italic:

Fixed.

> The 'more' drop down (or rather drop 'up') is cut of:

This is simply within the wrong container. I moved it up to `#content` so we can let it keep its opening position.

> Empty drop downs do not have the '-' (dash) displayed but have an empty option although when opening the drop down, the dash is selected:

~~This actually happens on production as well because angular doesn't properly `select` the given option. Not sure why, let's open a separate ticket for it.~~ Have found the reason even though outside this PR, fixed in https://github.com/opf/openproject/pull/5753/commits/cdf149712f323e3e4d4b269c5417788dc51bc7cd

> When using inline create, after every entry into a field, all input fields are redraw which leads to them visually disappearing for some time.

Fixed in the sense that they still need to redraw (think changing a type -> content of other fields such as status need to change), but the refreshing no longer replaces the DOM node.

> Switching type on wp create will sometimes cause a lot of js errors being reported in the console:

I could reproduce this initially, but can't anymore. What type are you switching to (especially regarding CFs?)

oliverguenther committed:
Append details dropdown to content

oliverguenther committed:
Don't scope subscription in WP edit form to table

oliverguenther committed:
Improve accessibility of display fields

oliverguenther committed:
Fix selected null placeholder '-' option

oliverguenther committed:
Reject fields to be refreshed, but no longer in the schema

oliverguenther commented:
I've adressed the issue of switching types. This happens consistently for type changes with a large amount of CFs because the angular digest responsible for adding and removing fields did not succeed before a field is being refreshed with the new schema (that do not have the CFs active).

In the long haul, we'll probably want to get rid of angular to control in the single-view itself as well.

I also load the WPs form as soon as a work package edit form is created for it, because that signals an impending editing that will require it.

ulferts commented:
Good to merge from my point of view. The minor annoyance of having the subject cut off to the right can be handled separately.

![image](https://user-images.githubusercontent.com/617519/28354956-30e9b1ae-6c62-11e7-8545-3abc98a41018.png)

ulferts commented:
Thanks for the work @oliverguenther
